### ISSUE TITLE
[FEATURE] - Length/duration in browse Route panel

### ISSUE THREAD

**Issue by JimmyS83** [cpesch/RouteConverter#85]
Hello,

would be possible to show (optionally?) very usefull parameters length/distance in Browser window for route? Currently, there is hardcoded just name and creator columns.

Use case: I have XX routes in folders, all by me. So author is not imported for me at all, but I would welcome options above, to quick glance into distance/ (especially) length  parameters, when comparing / changing routes, without need to click to Convert panel and back in every route change :)

**Comment by cpesch** [comment]
Hi @JimmyS83,

everything is possible. But...

1. I don't want RouteConverter to download all files just to extract that tiny bit of information since it overloads the network.
2. I don't want to parse the information on the server when RouteConverter requests it since it overloads the server.

So one could
a. let RouteConverter pass the information to the server – doesn't work for already uploaded stuff.
b. let the server do the parsing – somehow calling Java from Django/Python with limited resources.

Not easy...

**Comment by JimmyS83** [comment]
Understood.. 

I am using file locally (.routeconverter/routes) in gpx "track" format, for me, it seems that online route calculation isnt called, when I click to the Convert panel.. I didnt dig deep, its just a feeling.. When I load kml, then definitivelly I noticing route calculation. Personally, I have  (higher) tens of tracks, polished, which I am tracking in folders within .routeconverter routes, they are already calculated, I believe. I dont use cloud at all (generally), bcoz its usually slower, so I am (r)syncing/unisoning my files locally, when more devices are needed. 

My typical use case is choose folder (geo region), then first rough filter is choose, what length of track I wanna ride (therefore clicking between tabs all the time, considering number of saved tracks), and finally, compare "finalist" and choose optimal route using map viewer and browsing them.

But I understand your concerns, if your UI is aligned local+online, certain restriction regarding online could be expected. It was just thought, as I can see these info are already present in other tab like "immediatelly" when I click..

**Comment by cpesch** [comment]
Currently the UI is the same for local+online, but this could be changed. When describing your use case:
* what does length precisely mean? Distance in kilometers or duration in minutes?
* is it beeline for tracks? and routed over the street network for routes?
* how many positions do your tracks and/or routes typically have? How large are the files?

**Comment by JimmyS83** [comment]
- Actually both values interested me, as it depends heavilly if the track went through a lot cities, or countryside
- Not sure if I understand, attaching picture how it looks
![image](https://github.com/cpesch/RouteConverter/assets/7437057/11e9b13a-0454-4745-b10c-b053c1035aa7)

- I checked randomly, for example for 153km / 3h:20m route is 2149 positions. My gpx files has usually 70kB - 800kB per trip/track, depending on their length.

**Comment by cpesch** [comment]
That's a track which is rendered and calculated as a beeline, i.e. positions are connected with a straight line and no routing engine is involved.

A route typically consists of fewer positions which a routing engine connects with positions on the street network.

**Comment by JimmyS83** [comment]
Yes, thats why I think routing isnt actually called (therefore no load against server(s)), and I would still very welcome distance/time infos instead author in browser :)

These gpx which I am using are used for compatibility with a lot of my devices/apps - they are final product for "my library" - which is so usefull in Routeconverter in browser panel, when I figure out my next trip. 

When I need to correct them, I am using multiple tools on multiple platforms (Routeconverter, Cruiser, Kurviger ..), usually via .kml or .kurviger formats, but eventually, they are exported as gpx, and if thats not new route/track, the gpx is replaced.

So, when i edit time to time route, I really dont mind tabs whatsoever. But, when it comes to "library" function - ie. Browse panel, then distance/time are missed very much, and clicking back and forth to Convert tab is not convienient.

I understand, that primary goal for Routeconverter is convert/change.. But when you have very nice browser feature, switching offline gpx is fast, map view nearby, supporting even google traffic.. I just adopt this tool for library purposes as well, simply because I didnt find any tool better for that.. And the including time+distance in browse panel for offline gpx (which doesnt need to call exhausting routing engine) would be great..

I hope that clarify my use case little bit more...

**Comment by JimmyS83** [comment]
One thing comes to my mind - when I read 
_I don't want RouteConverter to download all files just to extract_ 

I am fine with showing that info only for current track/route **after** its selected and shown on map, if that would help minimize risk to overload servers.. Just like Conversion Tab, it also show length / time of one current, choosen track....

**Comment by cpesch** [comment]
You're right. I could let RouteConverter display the length after it has downloaded the file for as long as RouteConverter is running without being stopped. Would that help?

**Comment by JimmyS83** [comment]
I believe so, if I understand it correctly :)

**Comment by JimmyS83** [comment]
> You're right. I could let RouteConverter display the length after it has downloaded the file for as long as RouteConverter is running without being stopped. Would that help?

Also I just realized, in my original text I miss the other crucial parameter for me when I am planning afterwork trip: time (duration) :)  Basically I am switching the (locally) stored routes, and compare Length+Time to choose which I will take that day.  And that info is only in Convertion tab, but I am choosing/compare stored routes from Browse Panel.

So I hope it would be possible show both Time/Length values  somewhere within Browser/Map view UI. Time is also important, as I deciding also if I will make it to home at time :)

**Comment by cpesch** [comment]
Good news — this is implemented now.

The Browse panel has optional **Length** and **Duration** columns (toggle them from the *View* menu). You can sort by either and the column choice is remembered between sessions.

It also covers your exact case: for **locally stored routes** RouteConverter computes length and duration in the background, so you can compare them across a folder without opening each route in the Convert tab. For routes from the online catalog the values are now parsed server-side and delivered straight to the Browse panel, so they no longer disappear when you restart.

It's in the current **prerelease**: https://www.routeconverter.com/releases/

Could you give it a try and let me know whether it works the way you need for picking a route before a trip? Any feedback on the columns or their behaviour is very welcome.

**Comment by JimmyS83** [comment]
> Could you give it a try and let me know whether it works the way you need for picking a route before a trip? Any feedback on the columns or their behaviour is very welcome.

Just returned from vacation, by quick glance it looks really good! It would be great if there would be way how to store calculated values for locally stored gpx'es (to not being calculated all again each instance run), but I am aware of format restriction, and possibility to need to diff each gpx for changes, so there must be stored somewhere in cache, take consideration with different routing services etc.. And I understand that calculating route is dynamic feature anyway :) 

This will improve QoL of mine by miles! Thank you!

**Comment by cpesch** [comment]
I don't get the point how storing the calculated values improves your QoL that much. I mean: you now **have** the numbers available. That's QoL. So help me to understand what the real issue behind your question is.

**Comment by JimmyS83** [comment]
I meant that current change improved QoL of mine by miles (will was meant like its not yet in release branch) :)

**Comment by cpesch** [comment]
Ok, understood 🎉🚀

**Comment by cpesch** [comment]
Implemented — length/duration now shown in the browse Routes panel (RoutesTableCellRenderer). Confirmed working in prerelease; ships in 3.6 stable.
